app container stats collect, part 1, config part

naiming-zededa opened the pull request:
Signed-off-by: Naiming Shen <naiming@zededa.com>
- first part of the app container stats collection
- config change trigger the launch of goroutine for stats collection the first time
- will shutdown if there is no more appinstance needs this.

eriknordmark commented on pkg/pillar/cmd/zedrouter/appcontainer.go:
```diff
@@ -0,0 +1,74 @@
+	if strings.Compare(config.GetStatsIPAddr, oldIPAddr) != 0 {
+		log.Infof("appCheckStatsCollect: config ip %s, status ip %s", config.GetStatsIPAddr, oldIPAddr)
+		if oldIPAddr == "" && config.GetStatsIPAddr != "" {
+			ctx.appStatsMutex.Lock()
+			publishAppNetworkStatus(ctx, status)
```
1. Why hold mutex over publish? 2. Shouldn't you publish status for all changes not just the endbale case?

eriknordmark commented on pkg/pillar/cmd/zedrouter/appcontainer.go:
```diff
@@ -0,0 +1,74 @@
+					// XXX temp for later, collection function to fill in here
+				}
+			}
+			appStatsCollectTimer = time.NewTimer(600 * time.Second)
```
Do we want this timer to be setable using ConfigItemMap? 10 minutes sounds very long as a default since we report other metrics every 30 seconds or so.

eriknordmark commented on pkg/pillar/cmd/zedrouter/appcontainer.go:
```diff
@@ -0,0 +1,74 @@
+	for {
+		select {
+		case <-appStatsCollectTimer.C:
+			var numStatsIP int
+			ctx.appStatsMutex.Lock()
+			pub := ctx.pubAppNetworkStatus
+			items := pub.GetAll()
+			for _, st := range items {
+				status := st.(types.AppNetworkStatus)
+				if status.GetStatsIPAddr != "" {
+					numStatsIP++
+				}
+			}
+			if numStatsIP == 0 {
+				log.Infof("appStatsCollect: no stats IP anymore. stop and exit out")
+				ctx.appCollectStatsRunning = false
+				ctx.appStatsMutex.Unlock()
+				return
+			}
+			ctx.appStatsMutex.Unlock()
```
Might be useful encapsulate this in a function stopIfNoGetStats(ctx) bool which hides the mutex and the Running bool.

eriknordmark commented on pkg/pillar/cmd/zedrouter/appcontainer.go:
```diff
@@ -0,0 +1,74 @@
+		if oldIPAddr == "" && config.GetStatsIPAddr != "" {
+			ctx.appStatsMutex.Lock()
+			publishAppNetworkStatus(ctx, status)
+			if !ctx.appCollectStatsRunning {
+				ctx.appStatsMutex.Unlock()
+				go appStatsCollect(ctx)
```
This can easily start multiple instances of the goroutine. I think you need to set appCollectStatsRunning under the mutex to avoid that.

Also, might want to encapsulate this under a ensureStatsCollectRunning(ctx) to hide the logic.

eriknordmark commented on pkg/pillar/types/zedmanagertypes.go:
```diff
@@ -69,6 +69,8 @@ type AppInstanceConfig struct {
 	CloudInitUserData *string // base64-encoded
 	RemoteConsole     bool
+	// Collect Stats IP Address
+	CollectStatsIPAddr string
```
Would it be easier to track this as a net.IP instead of string?
From the string definition it isn't clear to me whether it should/does include a TCP port number etc aka net.TCPAddr

naiming-zededa commented:
- added the ConfigItemMap for the stats-collect interval
- changed the code to address the comments, mainly on the locking mutex.
- changed string to net.IP

eriknordmark commented on pkg/pillar/types/global.go:
```diff
@@ -705,6 +707,7 @@ func NewConfigItemSpecMap() ConfigItemSpecMap {
 	configItemSpecMap.AddIntItem(NetworkTestTimeout, 15, 0, 0xFFFFFFFF)
 	configItemSpecMap.AddIntItem(NetworkSendTimeout, 120, 0, 0xFFFFFFFF)
 	configItemSpecMap.AddIntItem(Dom0MinDiskUsagePercent, 20, 20, 0xFFFFFFFF)
+	configItemSpecMap.AddIntItem(AppContainerStatsInterval, 300, 60, 0xFFFFFFFF)
```
That's a min of 60 seconds, and I suspect there are sitations where we'd want 5 seconds to be able to demo this better. So the min can be 1 second for all I know.

naiming-zededa commented:
updated the doc and min setting. wondering how do we set this type of config...

eriknordmark commented on docs/CONFIG-PROPERTIES.md:
```diff
@@ -29,6 +29,7 @@
 | debug.default.remote.loglevel | string | warning | min level sent to controller |
 | storage.dom0.disk.minusage.percent | integer percent | 20 | min. percent of persist partition reserved for dom0 |
 | storage.apps.ignore.disk.check | boolean | false | Ignore disk usage check for Apps. Allows apps to create images bigger than available disk|
+| timer.appcontainer.stats.interval | integer in seconds | 300 | collect applicatio container stats |
```
Typo "applicatio"

eriknordmark replied:
Please take care of this in a subsequent PR; I'll pull this one

eriknordmark commented:
> updated the doc and min setting. wondering how do we set this type of config...

See https://github.com/lf-edge/eve/blob/master/docs/README.md#runtime-configuration-properties
has an example for setting the ssh key.
You can add an example for setting a timer if you'd like.
(And I see that page has an occurance of an old "zcli device" command

rvs commented:
LGTM
